tool/thingtalk-dataset.js

```diff
@@ -0,0 +1,312 @@
+//
+// See COPYING for details
+"use strict";
+
+const stream = require('stream');
+const ProgressBar = require('progress');
+
+const ThingTalk = require('thingtalk');
+const Grammar = ThingTalk.Grammar;
+const Library = ThingTalk.Ast.Input.Library;
+const Dataset = ThingTalk.Ast.Statement.Dataset;
+const TokenizerService = require('../lib/tokenizer');
+const OpenCC = require('opencc');
+
+
+class ThingTalkDatasetCleaner extends stream.Transform {
+    constructor(options) {
+        super({ objectMode: true });
+        
+        if (options.keepKeys && options.dropKeys)
+            throw 'keepKeys and dropKeys cannnot be set at the same time.'
+        this.keepKeys = options.keepKeys;
+        this.dropKeys = options.dropKeys;
+        this.options = options;
+        
+        this._defaults = {
+            'id': -1,
+            'type': '',
+            'annotations': {},
+            'utterances': [],
+            'preprocessed': []
+        }
+    }
+
+    _transform(ex, encoding, callback) {
+        if (ex.type == 'example') {
+            let dropKeys = this.dropKeys;
+            if(this.keepKeys) {
+                dropKeys = Object.keys(ex.data).filter(key => !this.keepKeys.includes(key));
+            }
+            if (dropKeys)
+                dropKeys.forEach(key => ex.data[key] = this._defaults[key]);
+        }
+        callback(null, ex);
+    }
+
+    _flush(callback) {
+        process.nextTick(callback);
+    }
+}
+
+class ThingTalkDatasetPreprocessor extends stream.Transform {
+    constructor(options) {
+        super({ objectMode: true });
+        
+        this.locale = options.locale;
+        this.options = options;
+        
+        this._language = '';
+        this._tokenizer = TokenizerService.get(process.env.GENIE_USE_TOKENIZER, true);
+        this._num_examples = 0;
+        this._bar = null;
+        this._i = 0;
+        
+        if (this.locale.toLowerCase() == 'zh-tw') {
+            this._opencc_t2s = new OpenCC('t2s.json');
+            this._opencc_s2t = new OpenCC('s2t.json');
+        }
+    }
+
+    async _transform(ex, encoding, callback) {
+        if (ex.type == 'meta') {
+            this._language = ex.data.language;
+            this._bar = new ProgressBar('[ :bar ] :current/:total', { total: ex.data.num_examples });
+        } else if (ex.type == 'example') {
+            try {
+                ex.data.preprocessed = await this.preprocessAll(ex.data.utterances, ex.data.id);
+            } catch (e) {
+                console.log(ex.data.id);
+                console.log(ex.data.utterances);
+                console.log(ex.data.preprocessed);
+                throw e;
+            }
+            this._i += 1;
+            this._bar.tick({ current: this._i });
+        }
+        callback(null, ex);
+    }
+    
+    async preprocessAll(utterances, id) {
+        const promises = await utterances.map(async utterance => {
```

Comment:
By convention, our style is to just pass the result of `.map` to `Promise.all()` unless there is a need to modify the array. Also, you should not `await` the result of `.map`: it's an Array of Promise, not a Promise.